packages/liveblocks-core/src/lib/url.ts

```diff
@@ -71,3 +75,88 @@ export function url(
     (result, str, i) => result + encodeURIComponent(values[i - 1] ?? "") + str
   ) as URLSafeString;
 }
+
+/**
+ * Sanitize a URL (normalize www URLs, handle relative URLs, prevent XSS attacks, etc.)
+ *
+ * Accepted URLs:
+ * - Absolute URLs with an http or https protocol (e.g. https://liveblocks.io)
+ * - Absolute URLs with a `www` prefix (e.g. www.liveblocks.io)
+ * - Relative URLs (e.g. /path/to/page)
+ *
+ * The presence/absence of trailing slashes is preserved.
+ * Rejected URLs are returned as `null`.
+ */
+export function sanitizeUrl(url: string): string | null {
+  // If the URL starts with "www.", normalize it as an HTTPS URL
+  if (url.startsWith("www.")) {
+    url = "https://" + url;
+  }
+
+  try {
+    const isAbsolute = ABSOLUTE_URL_REGEX.test(url);
+    const urlObject = new URL(
+      url,
+      isAbsolute ? undefined : PLACEHOLDER_BASE_URL
+    );
+
+    if (urlObject.protocol !== "http:" && urlObject.protocol !== "https:") {
+      return null;
+    }
+
+    const hasTrailingSlash = TRAILING_SLASH_URL_REGEX.test(url);
+
+    // Instead of using URL.toString(), we rebuild the URL manually
+    // to preserve the presence/absence of trailing slashes.
+    const sanitizedUrl =
+      // 1. Origin, only for absolute URLs
+      (isAbsolute ? urlObject.origin : "") +
+      // 2. Pathname, with a trailing slash if the original URL had one
+      (urlObject.pathname === "/"
+        ? // 2.a. Domain-only URLs, they always have their pathname set to "/"
+          hasTrailingSlash
+          ? "/"
+          : ""
+        : // 2.b. URLs with a path
+          hasTrailingSlash && !urlObject.pathname.endsWith("/")
+          ? urlObject.pathname + "/"
+          : urlObject.pathname) +
+      // 3. Search params
+      urlObject.search +
+      // 4. Hash
+      urlObject.hash;
+
+    return sanitizedUrl !== "" ? sanitizedUrl : null;
+  } catch {
+    return null;
+  }
+}
+
+/**
+ * Construct a URL with optional parameters and hash.
+ */
+export function generateUrl(
```

Comment:
There's probably more de-duplication possible between this and other utils above (`urljoin`, `toURLSearchParams`, etc) but I wanted to stay conservative in this PR to avoid doing too many things at once.